src/dos/program_setver.cpp

```diff
@@ -0,0 +1,721 @@
+		return;
+	}
+
+	std::vector<std::string> params;
+	cmd->FillVector(params);
+
+	// Handle first parameter being a path to SETVER.EXE database
+	const bool is_database_candidate = !params.empty() &&
+	                                   IsNameWithPath(params[0]);
+	if (is_database_candidate && PreprocessName(params[0])) {
+		if (params[0].back() == '\\') {
+			if (params[0] != "Z:\\") {
+				WriteOut(MSG_Get("PROGRAM_SETVER_WRONG_TABLE"));
+				return;
+			}
+			params.erase(params.begin());
+		}
+	}
+
+	// Helper lambda for file name preprocessing
+	auto preprocess_name = [&](std::string& name, const bool permissive = false) {
+		if (!PreprocessName(name, permissive)) {
+			if (IsNameWithPath(name)) {
+				WriteOut(MSG_Get("SHELL_ILLEGAL_PATH"));
+			} else {
+				WriteOut(MSG_Get("SHELL_ILLEGAL_FILE_NAME"));
+			}
+			return false;
+		}
+		return true;
+	};
+
+	// Preprocess first parameter if name/path is expected
+	if (!params.empty() && !has_arg_global) {
+		// When removing file from the list, it is possible the disk it
+		// was located does not currently exist - so be permissive
+		const bool permissive = has_arg_delete;
+		if (!preprocess_name(params[0], permissive)) {
+			return;
+		}
+		// It shouldn't be a directory
+		if (!params[0].empty() && params[0].back() == '\\') {
+			WriteOut(MSG_Get("SHELL_EXPECTED_FILE_NOT_DIR"));
+			return;
+		}
+	}
+
+	// Detect illegal switch combinations
+	if ((has_arg_batch || has_arg_paged) && (has_arg_delete || has_arg_global)) {
+		WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH_COMBO"));
+		return;
+	}
+
+	// Handle cases with no file parameters
+	if (params.empty()) {
+		if (has_arg_delete && has_arg_global) {
+			CommandDeleteGlobalOnly(has_arg_quiet);
+		} else if (has_arg_delete) {
+			CommandDeleteAll(has_arg_quiet);
+		} else if (has_arg_global) {
+			WriteOut(MSG_Get("SHELL_MISSING_PARAMETER"));
+		} else {
+			CommandPrintAll(has_arg_batch, has_arg_paged);
+		}
+		return;
+	}
+
+	// From now on at least 1 parameter is guaranteed to exist
+
+	// Detect illegal switch combinations
+	if (has_arg_delete && has_arg_global) {
+		WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH_COMBO"));
+		return;
+	}
+
+	// Hande entry deletion
+	if (has_arg_delete) {
+		if (params.size() > 1) {
+			WriteOut(MSG_Get("SHELL_TOO_MANY_PARAMETERS"));
+		} else {
+			CommandDeletePerFile(params[0], has_arg_quiet);
+		}
+		return;
+	}
+
+	// Hande global version set
+	if (has_arg_global) {
+		if (params.size() > 1) {
+			WriteOut(MSG_Get("SHELL_TOO_MANY_PARAMETERS"));
+		} else {
+			CommandSet({}, params[0], has_arg_quiet);
+		}
+		return;
+	}
+
+	// Handle per-file version set
+	if (has_arg_batch || has_arg_paged) {
+		WriteOut(MSG_Get("SHELL_WRONG_SYNTAX"));
+	} else if (params.size() > 2) {
+		WriteOut(MSG_Get("SHELL_TOO_MANY_PARAMETERS"));
+	} else {
+		CommandSet(params[0], params[1], has_arg_quiet);
+	}
+}
+
+bool SETVER::ParseVersion(const std::string& version_str, FakeVersion& version)
```

Comment:
I strongly suggest rewriting this function using https://en.cppreference.com/w/cpp/regex

Reply:
Done. The last time I tried to use regular expressions in DOSBox Staging reviewer asked me to migrate away from them. We'll see...